### ISSUE TITLE
No 'responseText' property in response object on 403 (forbidden) error

### ISSUE THREAD

**Issue by lapotchkin** [oracle/cordova-plugin-wkwebview-file-xhr#10]
Hi!

I added authentication in my phonegap app. But when I send wrong password, I receive only properties 'status', 'statusText', 'readyState'. But I debugged your code and found that the plugin had 'responseText' property inside. It's very important to me to have 'responseText', because this properti is used to send to the application some messages like 'wrong credentials' or some other.

This object I found when I was debugging:
```json
{
  "_context": {
    "delegate": null,
    "requestHeaders": {
      "content-type": "application/json",
      "accept": "application/json, text/javascript, */*; q=0.01"
    },
    "responseHeaders": {
      "date": "Tue, 31 Oct 2017 16:25:21 GMT",
      "keep-alive": "timeout=65, max=499",
      "content-length": "302",
      "connection": "Keep-Alive",
      "content-type": "application/json; charset=utf-8",
      "server": "Apache/2.4.18 (Ubuntu)"
    },
    "listeners": {},
    "readyState": 4,
    "responseType": "text",
    "withCredentials": true,
    "upload": {
      "_context": {
        "listeners": {}
      }
    },
    "status": 403,
    "method": "GET",
    "url": "https://example.com/auth",
    "async": true,
    "responseURL": "http://example.com/auth",
    "statusText": "forbidden",
    "response": "{\"meta\":{\"errors\":[{\"code\":102,\"message\":\"Invalid password.\"}]}",
    "responseText": "{\"meta\":{\"errors\":[{\"code\":102,\"message\":\"Invalid password.\"}]"
  }
}
```

This properties I received in jqXHR object:
```json
{"readyState":4,"status":403,"statusText":"forbidden"}
```

**Comment by gvanmat** [comment]
The hxr.responseText property should return the value you have shown above.  The property returns the context value you have shown above with corresponding names.  What does your listener logic look like?

**Comment by lapotchkin** [comment]
My code for requests:

```javascript
    var _options = {
        type: options.type || 'get',
        url: options.url || undefined,
        contentType: 'application/json',
        dataType: 'json',
        crossDomain: true,
        cache: false,
        xhrFields: {
            withCredentials: true
        },
        timeout: options.timeout || 0,
        data: options.data || {},
        success: _onSuccessRequest,
        error: _onErrorRequest
    };

    /**
     * Send request
     */
    this.send = function () {
        try {
            $.ajax(_options);
        } catch (e) {
            Tools.showError(e);
        }
    };

    function _onErrorRequest(jqXHR, textStatus) {
        console.log(jqXHR);
    }
```

Console output:

```json
{"readyState":4,"status":403,"statusText":"forbidden"}
```

When I was debugging your plugin, I found that property was not defined in this code:

```javascript
  // define readonly properties.
  ["readyState", "response", "responseText", "responseURL", "responseXML",
    "status", "statusText", "upload"].forEach(function (propName)
  {
    Object.defineProperty(window.XMLHttpRequest.prototype, propName,
      {
        "get": function ()
        {
          console.log(propName + ' = ' +  this._context[propName]);
          return this._context[propName];
        }
      });
  });
```

In console I found only status properties for status 403. When I was making good request with status, 200, I found property 'responseText' in my console.

P.S.: Sorry for my English.

**Comment by lapotchkin** [comment]
I rewrote my own method 'send' and now it works fine:

```javascript
this.send = function () {
        var xhr, url;
        var methodsWithBody = ['put', 'post'];

        try {
            _checkUrl();

            xhr = new XMLHttpRequest();
            url = methodsWithBody.indexOf(_options.type) === -1
                ? _options.url + '?' + $.param(_options.data)
                : _options.url;

            xhr.addEventListener("loadend", onLoadEnd);
            xhr.open(_options.type, url);
            xhr.responseType = _options.dataType;

            if (methodsWithBody.indexOf(_options.type) === -1) {
                xhr.send();
            } else {
                xhr.setRequestHeader("Content-Type", "application/json");
                xhr.setRequestHeader("Accept", "application/json");
                xhr.send(_options.data);
            }
        } catch (e) {
            Tools.showError(e);
        }

        function onLoadEnd(event) {
            var jqXHR;

            try {
                jqXHR = getJqXhr(this._context);
                Tools.log('REQUEST LOADEND');
                if (jqXHR.status >= 400) {
                    Tools.log('REQUEST ERROR');
                    Tools.log(url);
                    _onErrorRequest(jqXHR, jqXHR.statusText);
                } else {
                    Tools.log('REQUEST SUCCESS');
                    _onSuccessRequest(jqXHR.responseJSON, jqXHR.statusText, jqXHR);
                }
                Tools.log('REQUEST COMPLETE');
            } catch (e) {
                Tools.showError(e);
            }
        }

        function getJqXhr(data) {
            return {
                readyState: data.readyState,
                status: data.status,
                statusText: data.statusText,
                responseText: JSON.stringify(data.response),
                responseJSON: data.response,
                getResponseHeader: function (name) {
                    return data.responseHeaders[name.toLowerCase()];
                }
            };
        }
    };
```

But You have an error in README.MD. You have to add 'new' for this object creation:

```javascript
//wrong
xhr = XMLHttpRequest();

//correct
xhr = new XMLHttpRequest();
```

So, I think it was my mistake. Sorry))))

**Comment by gvanmat** [comment]
I think this is the same issue reported in #9.  The onerror event is raised when it should be causing the xhr wrapper to remove some of the returned properties.

I just put in a fix.  Please uninstall and re-install the plugin from the master to verify the fix:
cordova plugin remove cordova-plugin-wkwebview-file-xhr
cordova plugin add https://github.com/oracle/cordova-plugin-wkwebview-file-xhr.git

**Comment by lapotchkin** [comment]
I uninstalled and re-installed the plugin and it works. I received responseText in my 'error' callback. Thank You very much.

**Comment by gvanmat** [comment]
Great news. This fix will be in the 2.1.0 release which will have several others including iOS 11 support.

**Comment by gvanmat** [comment]
Closing as a duplicate of issue #9.
